Adding a Keycloak High Availability section to Keycloak's docs

ahus1 opened the pull request:
The content was moved over from the Keycloak Benchmark subproject.

There is another PR https://github.com/keycloak/keycloak-web/pull/433 which contains necessary enhancements needed for this change.

Closes #24844

<!---
Please read https://github.com/keycloak/keycloak/blob/main/CONTRIBUTING.md and follow these guidelines when contributing to Keycloak
-->

ahus1 committed:
Adding a Keycloak High Availability section to Keycloak's docs

The content was moved over from the Keycloak Benchmark subproject.

Closes #24844

Co-authored-by: Pedro Ruivo <pruivo@redhat.com>
Co-authored-by: Michal Hajas <mhajas@redhat.com>
Co-authored-by: Kamesh Akella <kakella@redhat.com>
Co-authored-by: Ryan Emerson <remerson@redhat.com>
Co-authored-by: Anna Manukyan <amanukya@redhat.com>
Signed-off-by: Alexander Schwartz <aschwart@redhat.com>

ahus1 commented:
Hi @andymunro, 

this PR is is now ready for a review. As discussed before, I'm looking for a focused, possibly time-boxed review to catch the most obvious mistakes, so it is good enough for upstream. 

We can then include it in the KC23 release and receive feedback from the community. We could have another review after it has been merge and after Thanksgiving to raise the contents quality to be ready for downstream.

As this PR adds quite an amount of text, review this in GitHub's UI might be efficient. Therefore, see below on how to check out the PR and build it locally. As agreed before, here's also a ZIP file with a pre-build HTML view: [high-availability.zip](https://github.com/keycloak/keycloak/files/13435587/high-availability.zip)

Once you've completed your review locally, commit them and push them to a branch in your `origin` repository and let me know the branch name you used in a comment. I can then merge those changes into this PR, or we can create the PR together. 

There is another PR https://github.com/keycloak/keycloak-web/pull/433 which will builds a full preview of the keycloak.org website. I'll update that one shortly with instructions on how to build a local preview.

---

To check out this PR locally in a branch named `pr-24913-review`, assuming that the Keycloak upstream repository is named `upstream`.

```
git fetch upstream pull/24913/head:pr-24913-review
git checkout pr-24913-review
```

You're free to adjust the name `pr-24913-review` to match your local conventions and change it both commands above.

To build the new guide locally, run once to build all dependencies: 

```
mvn clean install -DskipTests
```

Afterwards, open the file `docs/guides/target/generated-docs/high-availability/index.html` in a local browser for previewing all guides of the HA group.

After changes in the `docs/guide` folder, run the following command to incrementally the guides, which will be a lot faster than the command above: 

```
mvn clean install -pl docs/guides
``` 

A more general guide is available in `docs/guides/GENERATE-DOCS.md`

vmuzikar reviewed (commented):
@ahus1 Thanks for the PR. It's good to see the "scalability blueprint" gets more visibility.

Just out of curiosity, do we really want to cover AWS specific steps in official KC guides? I realize it's marked as unsupported preview but it still implies we intend to support it one day. That means we'll have to test the scenarios and maintain it.

ahus1 commented:
@vmuzikar - For now I just assume those guides are something that people want to see from Keycloak, and that they expect that Keycloak works on AWS.

If we see people actually using them, we should eventually support them. At the same time, not all guides listed here might change to supported at the same time, and the blueprints receive a different disclaimer once they graduate from the preview status.

Previous versions of the Cross-DC guide were using VMs to set things up, and at the same time left out bits around load balancer and the database (if I recall correctly).

With the Keycloak Benchmark project, we want to show one way of an actual deployment, and also figure out where Keycloak doesn't work as expected, and where we need to simplify the process. The manual steps here are automated in the KCB project. At the same time we're not using a shell script or Helm chart here (yet) as those in KCB contain extra bits and pieces, and as they don't explain the concepts so much and make it difficult to understand why we did things the way we did them. 

I assume people would use the blueprints to adapt them, and not many people would keep them as is. Still it is a lot simpler to start from something that works, than trying to figure things from scratch or from generic manuals.

In order to make them fully supported (and therefore also tested), maybe the blueprints need to adapt to use a cloud formation script for the AWS parts to have things more automated. We need to find the trade off to readability once we would do such changes, so let's see.

vmuzikar commented:
@ahus1 Thanks for the context.

Should we maybe add some disclaimers in the sense of "this is just a suggested approach, you need to adapt it for your environment, plus it's something that we can't guarantee it works". Of course, worded differently. :)

thomasdarimont commented on docs/guides/high-availability/concepts-memory-and-cpu-sizing.adoc:
```diff
@@ -0,0 +1,95 @@
+This ensures a fast startup of the node, and sufficient capacity to handle failover tasks like, for example, re-balancing Infinispan caches, when one node fails.
+Performance of {project_name} dropped significantly when its Pods were throttled in our tests.
+
+=== Calculation example
```
In my experience, the numbers used in the calculation show a low to medium utilization scenario.
This is probably the scenario that most (~80%) Keycloak users will encounter.

However, 10-20% of the Keycloak projects I have seen have a high to very high utilization scenario, e.g. 1,000,000 - 10,000,000 concurrent user sessions (with at least one client).

In the previous explanation it is written that CPU and memory scale linearly. Nevertheless, I suggest to explicitly explain a high load scenario with 5,000,000 user sessions.

thomasdarimont replied:
In addition, the user sessions of users do not have to remain in RAM forever.
In the case of offline user/client sessions, the memory requirement can also be reduced somewhat over time by shortening the cache entry lifetime. 
https://github.com/keycloak/keycloak/discussions/24274

ahus1 replied:
@thomasdarimont for your two comments, please create two new issues in the KCB project for this: https://github.com/keycloak/keycloak-benchmark/issues

We could over there collaborate on this and track the progress.

thomasdarimont reviewed (changes requested):
Hi Alex,

this comprehensive guide will be super useful for users, thank you very much for putting this together!

I read through the current state of the guide and hand a few remarks and suggestions. Feel free to adapt as you like.

ahus1 committed:
Updated the preview banner

Signed-off-by: Alexander Schwartz <aschwart@redhat.com>

thomasdarimont commented:
I think the guide could also address the topic of sticky sessions as they can also have an impact on the overall performance. The Keycloak Book (2nd) Edition recommends on page 167  (Keeping session affinity) to configure Keycloak to rely on the session affinity provided by the (reverse) proxy via the `spi-sticky-session-encoder-infinispan-should-attach-route=false` . This make sense when an embedded infinispan is used. However if keycloak is configured to use an external infinispan cluster. The recommendation from the book apparently doesn't apply anymore, becose the session "state" no longer resides in Keycloak. However, one could argue that the "local" caches still might contain some user data that is not loaded on other ("untouched" instances here).

ahus1 committed:
Reduce the visibility of tiles which shouldn't be visible on the top level

Signed-off-by: Alexander Schwartz <aschwart@redhat.com>

ahus1 commented:
@vmuzikar: Here's a suggestion which tries to keep the bar high for providing something useful, while pointing out there is more work to do:

> We provide these blueprints to show a minimal functionally complete example with a good baseline performance for regular installations.
You would still need to adapt it to your environment and your organization's standards and security best practices.

ahus1 commented:
> This make sense when an embedded infinispan is used. However if keycloak is configured to use an external infinispan cluster. The recommendation from the book apparently doesn't apply anymore, because the session "state" no longer resides in Keycloak.

@thomasdarimont - thank you for bringing this up. When using an external Infinispan, the data is now in two places: the embedded Infinispan and the external Infinispan. Going deeper into the code for user sessions Keycloak actually needs to have them all in the embedded Infinispan to be able to list all sessions of a user. 

As fetching information from the embedded local Infinispan is faster than fetching it from the remote store or a neighbor node, I say this options should be enabled in all cases. Looking at this code

https://github.com/keycloak/keycloak/blob/8a0fa521c45c175955bbe93be39efa9f5850f2d6/model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/util/InfinispanKeyGenerator.java#L61

It seems to depend on that option to attach the route to the cookie, which is hard for me to understand. I'll create a new issue in KCB for this to investigate more. Happy to jump on a 1:1 chat if you want.

ahus1 committed:
Update docs/guides/high-availability/connect-keycloak-to-external-infinispan.adoc

Co-authored-by: Thomas Darimont <thomas.darimont@googlemail.com>

ahus1 committed:
Update docs/guides/high-availability/concepts-threads.adoc

Co-authored-by: Thomas Darimont <thomas.darimont@googlemail.com>

vmuzikar commented:
> Here's a suggestion which tries to keep the bar high for providing something useful, while pointing out there is more work to do:

Makes sense, thanks.

ahus1 committed:
Adding disclaimer for blueprints

Signed-off-by: Alexander Schwartz <aschwart@redhat.com>

ahus1 commented:
@vmuzikar - added to the blueprint overview at the top, and to the two AWS specific guides (Aurora and Route53 loadbalancer).

ahus1 commented:
@thomasdarimont - ok, reading it again, as the first node would add the auth cookie which would then instruct the load balancer to redirect to the real node which would  have the session. Yes, disabling that would make sense in this setup as our loadbalancer doesn't check any cookies. It is an optimization, and I'll add it to our list of things to do.

ahus1 commented:
> I also tried to build keycloak-web with these changes to see the final look but I'm getting following exception:

@vmuzikar to build this PR with keycloak-web, there is another PR to keycloak-web to make this work as the previous version didn't handling subdirectories in the image folder. See https://github.com/keycloak/keycloak-web/pull/433

This was hidden in one of the issues in this conversation, I've now added it to the summary on the top. Sorry for now doing this earlier. 
Note that after the initial review we found putting all guides with a tile was quite overloaded, so I've added a change to put only some of the guides with tiles to the overview. The "hidden" guides are only linked from other guides, especially the first overview guide.

andymunro committed:
Edit HA documentation

Closes #pr-24913-review

andymunro committed:
Cleanup

Signed-off-by: Alexander Schwartz <aschwart@redhat.com>

stianst commented on docs/guides/high-availability/operate-failover.adoc:
```diff
@@ -0,0 +1,32 @@
+
+A failover from the primary site to the secondary site will happen automatically based on the checks configured in the loadbalancer.
+
+When the primary site loses its state in {ispn} or a network partition occurs that prevents the synchronization, manual procedures are necessary to recover the primary site before it can handle traffic again, see the <@links.ha id="operate-switch-back" /> {section}.
```
{ispn} shown in rendered output:
![image](https://github.com/keycloak/keycloak/assets/2271511/d606b905-1af7-4a68-99e8-da603b67458a)

ahus1 replied:
fixed in next commit

stianst commented on docs/guides/high-availability/operate-switch-over.adoc:
```diff
@@ -0,0 +1,91 @@
+
+include::partials/infinispan/infinispan-cli-connect.adoc[]
+
+. Disable the replication to the primary site by running the following command:
++
+Command:
+[source,bash,subs="+attributes"]
+----
+site take-offline --all-caches --site={site-a-cr}
+----
++
+Output:
+[source,bash,subs="+attributes"]
```
Command/Output not rendered in bold, and there's a `+` shown before Output:
![image](https://github.com/keycloak/keycloak/assets/2271511/122d5e30-8244-4017-9147-5020ff44d86c)

ahus1 replied:
Thanks, I missed the leading dot to make "Command" and "Output" a title.

stianst reviewed (commented):
Overall this is very good. Found a few rendering issues, and typos, that should be addressed.

In the release notes I think it would be good to focus on the A/P capability, rather than the guide. Something like (not an actual proposal, just to illustrate what I mean): 

```
Preview support for active-passive deployments

A lot of work has gone into testing and verifying Keycloak for active-passive deployments to bring increased availability to deployments. Alongside this capability there is a comprehensive blueprint to allow successful deployment of Keycloak in this configuration.
```

Another thing (probably more nitpick) is the blueprints are not listed as separate guides under `guides.html`. I assumed they would be, to make it easier to navigate to the bits and pieces separately. I'm okay with leaving it like that, but then it's inconsistent with the operational procedures that are listed as separate guides. I think we should either list everything as separate guides, or not list the individual operational procedures.

Finally, perhaps the `Concepts for active-passive deployments` should be listed before `Multi-site deployments`? I presume in most cases folks should read that one first?

ahus1 committed:
Update docs/documentation/release_notes/topics/23_0_0.adoc

Co-authored-by: Stian Thorgersen <stian@redhat.com>

ahus1 committed:
Update docs/guides/high-availability/introduction.adoc

Co-authored-by: Stian Thorgersen <stian@redhat.com>

ahus1 committed:
Update docs/guides/high-availability/introduction.adoc

Co-authored-by: Stian Thorgersen <stian@redhat.com>

ahus1 commented on docs/guides/high-availability/introduction.adoc:
```diff
@@ -0,0 +1,43 @@
+preview="true" >
+
+{project_name} supports deployments that consist of multiple Keycloak instances that connect to each other using its embedded Infinispan; load balancers can distribute the load evenly across those instances.
+Those setups are intended for a transparent network with ultra-low latencies.
+They also do not assume any network topology of the deployed instances.
```
Will replace it in the next commit with "Those setups are intended for a transparent network on a single site."

ahus1 commented on docs/guides/high-availability/concepts-active-passive-sync.adoc:
```diff
@@ -0,0 +1,182 @@
+| Seconds to minutes (depending on the database)
+
+| Primary site
+| If none of the {project_name} nodes are available, the loadbalancer will detect the outage and redirect the traffic to the secondary partials/infinispan/.
```
It seems some search-and-replace went wild here. Replacing it with "site" in the next commit.

ahus1 committed:
Update docs/guides/high-availability/concepts-active-passive-sync.adoc

Co-authored-by: Stian Thorgersen <stian@redhat.com>

ahus1 committed:
Update docs/guides/high-availability/concepts-active-passive-sync.adoc

Co-authored-by: Stian Thorgersen <stian@redhat.com>

ahus1 committed:
Update docs/guides/high-availability/operate-switch-over.adoc

Co-authored-by: Stian Thorgersen <stian@redhat.com>

ahus1 commented on docs/guides/high-availability/operate-switch-over.adoc:
```diff
@@ -0,0 +1,91 @@
+  "status" : "offline"
+}
+----
++
```
This will be fixed with the dot added above.

ahus1 committed:
Update docs/guides/high-availability/operate-network-partition-recovery.adoc

Co-authored-by: Stian Thorgersen <stian@redhat.com>

ahus1 commented:
> The blueprints are not listed as separate guides under `guides.html`. I assumed they would be, to make it easier to navigate to the bits and pieces separately. I'm okay with leaving it like that, but then it's inconsistent with the operational procedures that are listed as separate guides. I think we should either list everything as separate guides, or not list the individual operational procedures.

When reviewing it within the team, we found that those blueprint titles like "Deploy AWS Aurora in multiple availability zones" where not clearly connected to the overall heading, and jumping right into that topic would confuse users. Therefore they are linked only from the blueprint overview and the introduction guide.

Having the operational guides at the top level is IMHO different: Once you have deployed such a a setup, you would jump into those directly.

> Finally, perhaps the `Concepts for active-passive deployments` should be listed before `Multi-site deployments`? I presume in most cases folks should read that one first?

The `Multi-site deployments` serves both as an introduction for the topic and an overview of all guides (even those not listed with tiles). That's why it is listed first, and should probably be read first.

ahus1 committed:
Final touches

Signed-off-by: Alexander Schwartz <aschwart@redhat.com>

ahus1 committed:
Update docs/guides/high-availability/concepts-active-passive-sync.adoc

Co-authored-by: Stian Thorgersen <stian@redhat.com>

ahus1 committed:
Switch to different include to enable freemarker in included file

Signed-off-by: Alexander Schwartz <aschwart@redhat.com>

vmuzikar commented:
> to build this PR with keycloak-web, there is another PR to keycloak-web to make this work as the previous version didn't handling subdirectories in the image folder. See [keycloak/keycloak-web#433](https://github.com/keycloak/keycloak-web/pull/433)

@ahus1 Thank you, missed that PR.

Did a high level review and LGTM. Nice stuff, we were missing those a long time!

ahus1 committed:
Updating the release notes

Signed-off-by: Alexander Schwartz <aschwart@redhat.com>

ahus1 commented:
@stianst - updated the release notes as follows: 

> **Preview support for multi-site active-passive deployments**
 
> Deploying Keycloak to multiple independent sites is essential for some environments to provide high availability and a speedy recovery from failures.
This release adds preview-support for active-passive deployments for Keycloak.

> A lot of work has gone into testing and verifying a setup which can sustain load and recover from the failure scenarios.
To get started, use the high-availability guide which also includes a comprehensive blueprint to deploy a highly available Keycloak to a cloud environment.

Final question: I noticed that the [enable/disable feature guide](https://www.keycloak.org/server/features) will never list cross-site as a feature, as it is nothing you enable on the CLI. Would we need a central location to state that this is preview? Should we add it to that page manually?

ahus1 commented on docs/documentation/release_notes/topics/23_0_0.adoc:
```diff
@@ -83,3 +83,15 @@ Keycloak supports new password policy, which allows to specify the maximum age o
 When this password policy is set to 0, the user will be required to re-authenticate to change the  password in the Account Console or by other means.
 You can also specify a lower or higher value than the default value of 5 minutes. Thanks to https://github.com/thomasdarimont[Thomas Darimont] for the contribution.
 
+= New {project_name} high availability guide in preview
```
updated in last commit.

ahus1 committed:
Update docs/guides/high-availability/concepts-active-passive-sync.adoc

Co-authored-by: Stian Thorgersen <stian@redhat.com>

ahus1 committed:
Update docs/guides/high-availability/concepts-active-passive-sync.adoc

Co-authored-by: Stian Thorgersen <stian@redhat.com>

ahus1 commented on docs/guides/high-availability/partials/infinispan/infinispan-cli-connect.adoc:
```diff
@@ -0,0 +1,21 @@
+----
++
+It asks for the username and password for the {ispn} cluster.
+Those credentials are the one set in the <@links.ha id="deploy-infinispan-kubernetes-cross-dc"/> {section} in the configuring credentials section.
```
Fixed in last commit.

ahus1 committed:
Update docs/guides/high-availability/concepts-threads.adoc

Co-authored-by: Stian Thorgersen <stian@redhat.com>

ahus1 committed:
Update docs/guides/high-availability/concepts-active-passive-sync.adoc

Co-authored-by: Stian Thorgersen <stian@redhat.com>

ahus1 commented on docs/guides/high-availability/concepts-active-passive-sync.adoc:
```diff
@@ -0,0 +1,182 @@
+
+Is a synchronous cluster less stable than an asynchronous cluster?::
+An asynchronous setup would handle network failures between the site gracefully, while the synchronous setup would delay requests and will throw errors to the caller where the asynchronous setup would have deferred the writes to {jdgserver_name} or the database to the secondary site.
+However, as the secondary partials/infinispan/ would never be fully up-to-date with the primary partials/infinispan/, this setup could lead to data loss during failover.
```
Fixed, bad replace.

ahus1 committed:
Update docs/guides/high-availability/concepts-threads.adoc

Co-authored-by: Stian Thorgersen <stian@redhat.com>

ahus1 commented on docs/guides/images/high-availability/infinispan-crossdc-az.dio.svg:
Yes: Each site provides one GossipRouter, and they server as a HA-setup. Messages can go either route.

ahus1 committed:
Update docs/guides/high-availability/deploy-infinispan-kubernetes-crossdc.adoc

Co-authored-by: Stian Thorgersen <stian@redhat.com>

ahus1 commented on docs/guides/high-availability/bblocks-active-passive-sync.adoc:
```diff
@@ -0,0 +1,68 @@
+
+*Blueprint:* <@links.ha id="deploy-infinispan-kubernetes-crossdc" /> using the {jdgserver_name} Operator, and connect the two sites using {jdgserver_name}'s Gossip Router.
+
+*Not considered:* Direct interconnections between the Kubernetes clusters on the network layer.
```
Stretched clusters have been ruled out above already for other reasons. This is about interconnections like submariner, where you have independent clusters, and then start to connect them https://submariner.io/

ahus1 commented on docs/guides/high-availability/connect-keycloak-to-external-infinispan.adoc:
```diff
@@ -0,0 +1,79 @@
+This site name is related only to the embedded {jdgserver_name} and does not need to match any value from the external {jdgserver_name} deployment.
+<5> Mounting the cache configuration Volume in Kubernetes.
+However, matching the `jboss.site.name` with the external {jdgserver_name} deployment site name helps debugging possible future issues.
+If you are using multiple sites for {project_name} in a cross-DC setup such as <@links.ha id="deploy-infinispan-kubernetes-crossdc" />, the site name must be different in each site.
```
fixed in next commit

pedroigor commented on docs/guides/high-availability/concepts-active-passive-sync.adoc:
```diff
@@ -0,0 +1,174 @@
+This would include:
++
+--
+* Lost logouts, meaning sessions are logged in the secondary site although they are logged out in to the primary site at the point of failover when using an asynchronous {jdgserver_name} replication of sessions.
```
Emails containing links to flows like password reset should also not work. Is it worth mentioning here too?

ahus1 commented on docs/guides/high-availability/bblocks-active-passive-sync.adoc:
```diff
@@ -0,0 +1,68 @@
+
+* Understanding the concepts laid out in the <@links.ha id="concepts-active-passive-sync"/> {section}.
+
+== Two sites with low-latency connection
+
+Ensures that synchronous replication is available for both the database and the external {jdgserver_name}.
+
+*Blueprint:* Two AWS Availablity Zones within the same AWS Region.
+
+*Not considered:* Two regions on the same or different continents, as it would increase the latency and the likelihood of network failures.
+Synchronous replication of databases as a services with Aurora Regional Deployments on AWS is only available within the same region.
+
+== Environment for {project_name} and {jdgserver_name}
+
+Ensures that the instances are deployed and restarted as needed.
+
+*Blueprint:* Red Hat OpenShift Service on AWS (ROSA) deployed in each availability zone.
+
+*Not considered:* A stretched ROSA cluster which spans multiple availability zones, as this could be a single point of failure if misconfigured.
```
The next commit changes those two "Blueprint" to "Suggested setup" where we don't have blueprint. Maybe we would have a blueprint for ROSA in the next iteration.

ahus1 committed:
Final touches

Signed-off-by: Alexander Schwartz <aschwart@redhat.com>

pedroigor commented on docs/guides/high-availability/concepts-memory-and-cpu-sizing.adoc:
```diff
@@ -0,0 +1,96 @@
+
+* For each 40 user logins per second, 1 vCPU per Pod in a three-node cluster (tested with up to 300 per second).
++
+{project_name} spends most of the CPU time hashing the password provided by the user.
```
Are these recommendations based on the default hashing iterations? Does it make sense to mention this?

ahus1 replied:
This is described as part of "Reference architecture" below

pedroigor commented on docs/guides/high-availability/concepts-memory-and-cpu-sizing.adoc:
```diff
@@ -0,0 +1,96 @@
+
+=== Calculation example
+
+Target size:
```
Local caches such as `realm` and `users` should also impact the memory requested/limits. Does it make sense to give an idea about the number of realms, clients, and users the recommendation is based on?

pedroigor replied:
Ok, also in the following statements.

ahus1 replied:
This is using the defaults, and we currently don't have a recommendation to change them. The tests were running with more users than would fit into the cache, so Keycloak was loading them from the database (at least once per login). So we should be safe here.

pedroigor commented on docs/guides/high-availability/concepts-memory-and-cpu-sizing.adoc:
```diff
@@ -0,0 +1,96 @@
+* {project_name} deployed with the Operator and 3 pods.
+* Default user password hashing with PBKDF2 27,500 hash iterations.
+* Database seeded with 100,000 users and 100,000 clients.
+* Infinispan caches at default of 10,000 entries, so not all clients and users fit into the cache, and some requests will need to fetch the data from the database.
```
As stated here 10k max-entries is not enough to store all realm data. Based on the 100k users/clients don't we want to recommend a better limit for local caches? I don't know how to calculate this but ideally, we should keep the state in memory as much as possible to avoid unnecessary database roundtrips?

ahus1 replied:
In a real world scenario you wouldn't have all the users in memory anyway, and especially on restarting nodes those would have an empty cache at the beginning. 

Running a load test where Keycloak is loading data from the database continuously and succeeds gives me more confidence than a load test where all data is in memory all the time. The 10_000 cache entries are enough to ensure that a user is cached during the authentication flow, which is good enough and already reduces a lot of the load.

pedroigor replied:
I'm not 100% convinced about that based on the deployments I've seen. This is especially true when you have a load spike and failures that force users to re-authenticate.

For the realm `cache`, depending on the number of roles, groups, clients, and other related data, this number is also not enough.

For production deployments I don't think the right recommendation should be use the default values. But at least recommend trying out increasing these local caches to match the expected performance and database resource usage (IIOPS are expensive depending on your deployment and we had cases around that in the past).

Don't you think we should at least mention the points we are discussing here? For instance:

```
The default values for local caches such as `realm` and `users` caches should be adjusted accordingly to your performance requirements, database resource usage, and memory constraints. Make sure to run load tests as close as possible to your real load and monitor how resources are being used to better adjust these settings. 
```

thomasdarimont replied:
Btw. the local caches (https://github.com/keycloak/keycloak/blob/main/quarkus/runtime/src/main/resources/cache-ispn.xml#L26) in don't seem to have any expiration policy configured. I think it would be helpful to enable some expiration to allow keycloak to release more memory if the user isn't used for a while (e.g. 8h).

ahus1 replied:
> The local caches don't seem to have any expiration policy configured.

@thomasdarimont - this is not related to this guide, and probably a default that Keycloak should change. Please create an issue in Keycloak's main repository and submit a pull request with your ideas around that and discuss it. It could be 8 or 24 or 36 hours. Ping me and @pedroigor in the PR and we'll listen to your thoughts on this.

ahus1 replied:
> For production deployments I don't think the right recommendation should be use the default values. But at least recommend trying out increasing these local caches to match the expected performance and database resource usage (IIOPS are expensive depending on your deployment and we had cases around that in the past).

There is already a paragraph at the top with recommendations to increase the cache. Will add a hit towards IOPS there in https://github.com/keycloak/keycloak/issues/24975.

I think this is best tuned when looking at the metrics. For https://github.com/keycloak/keycloak-benchmark/issues/579, I've added "Cache hit ratio + cache size for local caches".

pedroigor commented on docs/guides/high-availability/concepts-memory-and-cpu-sizing.adoc:
```diff
@@ -0,0 +1,96 @@
+
+* For each 450 client credential grants per second, 1 vCPU per Pod in a three node cluster (tested with up to 2000 per second).
++
+Most CPU time goes into creating new TLS connections, as each client runs only a single request.
```
I also think we should recommend re-encrypting (from ingress) traffic to backend nodes. It is worth mentioning how TLS termination (and the different options like `edge`) can impact this specific usage of CPU?

ahus1 replied:
The test succeeded with both sustaining the load and recovering from failures. 

I see this as an optimizations, and maybe you could open an issue in the keycloak benchmark project to run such a test and measure the differences. The Keycloak operator doesn't provide a configuration for certmanager and/or OpenShift to do reencrypt, so once the test for a manually configured Keycloak shows a performance improvement, the next step would be a better Operator support, and then we could eventually update this guide (or the other way around, but it would be a bit tedious for customers)

pedroigor replied:
I don't want to put more stuff in `keycloak-benchmark` backlog if not needed. It is more about whether we want to recommend a more secure deployment (which does not really depend on us to decide) with re-encrypt.

Just wanted to highlight this here to see if it makes sense to include it now. Eve if we just want to state the numbers are based on a specific proxy mode.

ahus1 replied:
Ok, I'll add a sentence in the follow-up issue https://github.com/keycloak/keycloak/issues/24975

pedroigor commented on docs/guides/high-availability/concepts-memory-and-cpu-sizing.adoc:
```diff
@@ -0,0 +1,96 @@
++
+{project_name} spends most of the CPU time hashing the password provided by the user.
+
+* For each 450 client credential grants per second, 1 vCPU per Pod in a three node cluster (tested with up to 2000 per second).
```
Not sure if deserves a note about this but that is probably based on a client setting where refresh tokens are disabled (default). Some people might be using or might want (don't ask me why but sadly we used to support that) refresh tokens when doing client credentials and that might change the results a lot.

ahus1 replied:
Will add a sentence to the reference architecture in https://github.com/keycloak/keycloak/issues/24975

thomasdarimont commented on docs/guides/high-availability/connect-keycloak-to-external-infinispan.adoc:
```diff
@@ -0,0 +1,79 @@
+preview="true"
+tileVisible="false" >
+
+This topic describes advanced {jdgserver_name} configurations for {project_name} on Kubernetes.
```
One aspect that should also be considered when using an external Infinispan cluster is that the current readiness probes only check for a working database connection. However, an additional test for a working Infinispan connection would be useful here.

ahus1 replied:
We have the following issue for this: https://github.com/keycloak/keycloak-benchmark/issues/591 and hope to fix it soon. When you look for "Infinispan cluster failure" in "concepts-active-passive-sync.adoc" you'll see it listed there that human intervention is needed.

pedroigor commented on docs/guides/high-availability/concepts-threads.adoc:
```diff
@@ -0,0 +1,69 @@
+Depending on the available CPU cores, it can grow even larger.
+Threads are created as needed, and will end when no longer needed, so the system will scale up and down as needed.
+
+When the load and the number of threads increases, the bottleneck will usually be the database connections.
```
I would also say CPU because you will have more context switches/thread contention. Basically, more threads do not necessarily mean more processing power but it can be the opposite as you are pushing too much on the CPU. If you have enough cores available, that is fine. Otherwise, you should consider lowering the number of threads in favor of faster response times (less context switches/contention) and better throughput.

In addition to what is being said here I would also consider:

```
You should also consider adjusting the number of worker threads to avoid creating more than what the CPU limits set to nodes can handle. Ideally, you should start with a low limit of threads and adjust it accordingly to the target throughput and response time.
```

pedroigor commented on docs/guides/high-availability/concepts-threads.adoc:
```diff
@@ -0,0 +1,69 @@
+This will use additional memory in the Pod, can exhaust resources in the load balancers, and the requests will eventually time out on the client side without the client knowing if the request has been processed.
+To limit the number of queued requests in {project_name}, set an additional Quarkus configuration option.
+
+Configure `quarkus.thread-pool.queue-size` to specify a maximum queue length to allow for effective load shedding once this queue size is exceeded: {project_name} will return HTTP Status code 500 (server error).
```
Our load-shedding mechanism is/will return 503 instead, right? Not sure if `500` is the right status code...

ahus1 commented on docs/guides/high-availability/concepts-threads.adoc:
```diff
@@ -0,0 +1,69 @@
+[#load-shedding]
+=== Load Shedding
+
+By default, {project_name} will queue all incoming requests infinitely, even if the request processing stalls.
```
For now it should be enough. We're having some ideas with the Quarkus team to do this differently, let's see how this evolves. Resolving this as your last sentence implies.

pedroigor commented:
@ahus1 This is awesome stuff and something people have been expecting from us for a long, long time. It is great to see all those points available from our documentation so that we can adjust them accordingly based on the feedback we get.

Congrats to all involved, great stuff ...

ahus1 commented on docs/guides/high-availability/concepts-threads.adoc:
```diff
@@ -0,0 +1,69 @@
+Threads are created as needed, and will end when no longer needed, so the system will scale up and down as needed.
+
+When the load and the number of threads increases, the bottleneck will usually be the database connections.
+Once a request cannot acquire a database connection, it will fail with a message in the log like `Unable to acquire JDBC Connection`.
```
Follow-up issue: https://github.com/keycloak/keycloak-benchmark/issues/603

ahus1 replied:
Will try to address this in https://github.com/keycloak/keycloak/issues/24975
